tools/bridge/bridge/transfer_fetcher.py

```diff
@@ -1,25 +1,135 @@
 
+from gevent.queue import Queue
 from web3 import Web3
 
+MINIMAL_TOKEN_ABI = [
+    {
+        "anonymous": False,
+        "inputs": [
+            {"indexed": True, "name": "from", "type": "address"},
+            {"indexed": True, "name": "to", "type": "address"},
+            {"indexed": False, "name": "value", "type": "uint256"},
+        ],
+        "name": "Transfer",
+        "type": "event",
+    }
+]
+
+TRANSFER_EVENT_SIGNATURE_HASH = Web3.keccak(text="Transfer(address,address,uint256)")
+
+
+class TransferFetcher:
+    def __init__(
+        self,
+        *,
+        queue: Queue,
+        w3_foreign: Web3,
+        token_contract_address: str,
+        foreign_bridge_contract_address: str,
+        max_reorg_depth: int,
+        event_filter_fetch_limit: int,
+    ):
+        token_contract_code = w3_foreign.eth.getCode(token_contract_address)
+
+        if not token_contract_code:
+            raise ValueError(
+                f"The given token contract address {token_contract_address} "
+                "does not point to a contract!"
+            )
+
+        if TRANSFER_EVENT_SIGNATURE_HASH not in token_contract_code:
+            raise ValueError(
+                f"The token contract at the given address {token_contract_address}"
+                "does not have the (correct) 'Transfer' event!"
+            )
+
+        foreign_bridge_contract_code = w3_foreign.eth.getCode(
+            foreign_bridge_contract_address
+        )
+
+        if not foreign_bridge_contract_code:
+            raise ValueError(
+                f"The given foreign bridge contract address {foreign_bridge_contract_address} "
+                "does not point to a contract!"
+            )
+
+        self._logger = logging.getLogger("bridge.transfer_fetcher")
+        self._w3_foreign = w3_foreign
+        self._queue = queue
+        self._token_contract = w3_foreign.eth.contract(
+            address=token_contract_address, abi=MINIMAL_TOKEN_ABI
+        )
+        self._foreign_bridge_contract_address = foreign_bridge_contract_address
+        self._max_reorg_depth = max_reorg_depth
+        self._event_filter_fetch_limit = event_filter_fetch_limit
+        self._last_fetched_block_number = (
+            -1
+        )  # Fetching starts always one block after the last one.
+
+    def _fetch_token_transfer_events_in_range(
+        self, from_block_number: int, to_block_number: int
+    ) -> None:
+        from_block_number = max(from_block_number, 0)
+        to_block_number = min(to_block_number, self._w3_foreign.eth.blockNumber)
+
+        self._logger.debug(
+            "Fetch transfer events from block {from_block_number} to {to_block_number}."
+        )
+
+        assert from_block_number <= to_block_number, (
+            f"Can not fetch from a negative range of blocks "
+            f"({from_block_number} to {to_block_number})"
+        )
+
+        transfer_event_filter = self._token_contract.events.Transfer.createFilter(
```

Comment:
We should not use `createFilter`, it won't work on infura. I *think* this is the right method: https://web3py.readthedocs.io/en/stable/web3.eth.html#web3.eth.Eth.getLogs (not sure if there's one on the contract object as well).

Reply:
Yes this was the problem I think. Sry for the mistake. Took the contract to make the argument filter easy. I will try it out to make sure it use the correct RPC method. In the link docu they also state to create a new filter and remove it directly. I'll check this.
Btw: Do you know a nice utility function to extend a 20 byte address to a 32 byte string. Need it regarding the topic filters. Has not worked for me in the past for some reason.

Reply:
`eth_getLogs` is the right method. Please do not create any filters. They do not survive a restart of the (parity/geth) node.

Reply:
@schmir yes, this is definitely not correct. :+1: 
Do you probably know a such requested utility function by one of all these nice Ethreum related libraries?

Reply:
A utility function doing what? I've implemented decoding of the log entries returned by eth_getLogs in py-eth-index if that is what you need.

Reply:
I adjusted it to go for `web3.eth.getLogs`.
:heavy_check_mark: